### DISCUSSION TITLE
Apple provider always redirects to 'baseUrl' (NEXTAUTH_URL) instead of 'callbackUrl'

### DISCUSSION THREAD

**Discussion by yuzhakovvv** [nextauthjs/next-auth#1704] [Help]
When user uses _Apple_ provider he's redirected to `baseUrl` (`NEXTAUTH_URL`) instead of `callbackUrl` that was passed to `signIn()` method. _Google_ provider works as expected and redirects to `callbackUrl`.

Sign In is successful as app gets user data from Apple. `callbacks` work as expected. Custom `callbacks.redirect` returns correct `callbackUrl`.

What can affect `callbackUrl` to be changed to `baseUrl` (`NEXTAUTH_URL`) after `callbacks.redirect`? Maybe you can point a line in the source code?

**Comment by wescarr** [comment]
I am having the same issue. I believe this is because Apple redirects back to `/api/auth/callback/apple` with a `POST` request which does **NOT** send the `__Secure-next-auth.callback-url` cookie in the request. Other providers, like Google, do the callback with a `GET` request.

IMO using a cookie to store the `callbackUrl` is a mistake, as it also does not work for the email provider, especially if want to use a different browser/session from where you provided your email address.

**Reply by balazsorban44** [reply]
One workaround is to adjust the cookie options to be able to use the cookie, similar to as described here:

https://github.com/nextauthjs/next-auth/issues/1664#issuecomment-815285862

I can see why this is problematic. Not using cookies would mean that we should rely on some other mechanism for storing information. Currently I don't have a better idea, but open for suggestions! 🙂

**Reply by wescarr** [reply]
@balazsorban44 thanks for your response. I believe a common pattern is to add the callback url as a query param to the `redirect_uri` sent to OAuth providers. Granted, I am not an expert on the security implications of that approach. Do you have thoughts/feedback on this approach?
